src/codemagic/tools/keychain.py

```diff
@@ -245,24 +271,45 @@ def _generate_path(self):
 
     @cli.action('add-certificates',
                 KeychainArgument.CERTIFICATE_PATHS,
-                KeychainArgument.CERTIFICATE_PASSWORD)
-    def add_certificates(self,
-                         certificate_path_patterns: Sequence[pathlib.Path],
-                         certificate_password: Password = Password('')):
+                KeychainArgument.CERTIFICATE_PASSWORD,
+                KeychainArgument.ALLOWED_APPLICATIONS,
+                KeychainArgument.ALLOW_ALL_APPLICATIONS,
+                KeychainArgument.DISALLOW_ALL_APPLICATIONS)
+    def add_certificates(
+            self,
+            certificate_path_patterns: Sequence[pathlib.Path],
+            certificate_password: Password = Password(''),
+            allowed_applications: Sequence[pathlib.Path] = KeychainArgument.ALLOWED_APPLICATIONS.get_default(),
+            allow_all_applications: bool = KeychainArgument.ALLOW_ALL_APPLICATIONS.get_default(),
+            disallow_all_applications: bool = KeychainArgument.DISALLOW_ALL_APPLICATIONS.get_default()):
         """
         Add p12 certificate to specified keychain.
         """
 
-        self.logger.info(f'Add certificates to keychain {self.path}')
+        add_for_all_apps = False
+        add_for_apps: Sequence[pathlib.Path] = []
+        if allow_all_applications and disallow_all_applications:
+            raise KeychainArgument.ALLOW_ALL_APPLICATIONS.raise_argument_error(
+                f'Using mutually exclusive options '
+                f'{KeychainArgument.ALLOWED_APPLICATIONS.flag!r} and '
+                f'{KeychainArgument.DISALLOW_ALL_APPLICATIONS.flag!r}')
+        elif allow_all_applications:
```

Comment:
since previous `if` raises an exception, just `if` can be sufficient here

Reply:
I prefer to keep all application spefic validations in a single block since conceptually they fit nicely together. Recent changes also make it slightly easier to read and follow.